File: src/camera_cus/camera_cus/camera_node.py
###
### Before start Node in Real:
### export LD_PRELOAD=/usr/lib/x86_64-linux-gnu/libunwind.so.8
###
import rclpy
from rclpy.node import Node
from sensor_msgs.msg import Image
from cv_bridge import CvBridge
import cv2
import numpy as np
import math
from nav_msgs.msg import Odometry
from geometry_msgs.msg import TransformStamped
from tf2_ros import TransformBroadcaster
from rclpy.qos import QoSProfile, QoSDurabilityPolicy, QoSHistoryPolicy, QoSReliabilityPolicy


def quaternion_from_euler(ai, aj, ak):
    ai /= 2.0
    aj /= 2.0
    ak /= 2.0
    ci = math.cos(ai)
    si = math.sin(ai)
    cj = math.cos(aj)
    sj = math.sin(aj)
    ck = math.cos(ak)
    sk = math.sin(ak)
    cc = ci*ck
    cs = ci*sk
    sc = si*ck
    ss = si*sk

    q = np.empty((4, ))
    q[0] = cj*sc - sj*cs
    q[1] = cj*ss + sj*cc
    q[2] = cj*cs - sj*sc
    q[3] = cj*cc + sj*ss

    return q

class CusCamera(Node):
    def __init__(self):
        super().__init__('cus_camera')
        self.get_logger().info(f"OpenCV version: {cv2.__version__}")
        self.set_parameters([rclpy.parameter.Parameter('use_sim_time',rclpy.Parameter.Type.BOOL, True)])
        self.odom_pub = self.create_publisher(Odometry, 'camera_odom', 1000)

        qos_profile = QoSProfile(
            depth=2,
            reliability=QoSReliabilityPolicy.BEST_EFFORT,
            durability=QoSDurabilityPolicy.VOLATILE,
            history=QoSHistoryPolicy.KEEP_LAST
        )
        # self.timer = self.create_timer(0.1, self.timer_callback)
        self.subscription = self.create_subscription(
            Image,
            '/camera/image', 
            self.callback,
            qos_profile=qos_profile)
        self.tf_broadcaster = TransformBroadcaster(self)

        self.bridge = CvBridge()
        
        # REAL CAMERA INTRINSIC PARAMS
        self.K = np.array([[ 1.05477276e+03, -1.10039944e-01,  1.07993991e+03],
                        [ 0.00000000e+00,  1.05437012e+03,  7.33474904e+02],
                        [ 0.00000000e+00,  0.00000000e+00,  1.00000000e+00]])
        self.D = np.array([[-0.03337077], [-0.00093301], [-0.00130544], [ 0.0011649 ]])
        

        dictionary = cv2.aruco.getPredefinedDictionary(cv2.aruco.DICT_4X4_250)
        parameters = cv2.aruco.DetectorParameters()
        self.detector = cv2.aruco.ArucoDetector(dictionary, parameters)

        # Aruco positions in playground coordinates in format:
        # <Aruco ID>: <corners in playground coordinates: top left, top right, bottom right, bottom left> 
        Aruco = {
            20: np.array([[550, 550, 0], [650, 550, 0], [650, 650, 0], [550, 650, 0]], dtype=np.float32),
            21: np.array([[2350, 550, 0], [2450, 550, 0], [2450, 650, 0], [2350, 650, 0]], dtype=np.float32),
            22: np.array([[550, 1350, 0], [650, 1350, 0], [650, 1450, 0], [550, 1450, 0]], dtype=np.float32),
            23: np.array([[2350, 1350, 0], [2450, 1350, 0], [2450, 1450, 0], [2350, 1450, 0]], dtype=np.float32)
        }
        objPoints = np.fromiter(Aruco.values(), dtype=object)
        ids = np.fromiter(Aruco.keys(), dtype=float)
        # print(objPoints, ids)
        dictionary = cv2.aruco.getPredefinedDictionary(cv2.aruco.DICT_4X4_50)
        self.board = cv2.aruco.Board(objPoints, dictionary, ids)

        self.marker_length = 80
        self.object_corners = np.array([
            [[-self.marker_length//2,  self.marker_length//2, 0]],  # Левый верхний
            [[ self.marker_length//2,  self.marker_length//2, 0]],  # Правый верхний
            [[ self.marker_length//2, -self.marker_length//2, 0]],  # Правый нижний
            [[-self.marker_length//2, -self.marker_length//2, 0]]   # Левый нижний
        ], dtype=np.float32)

        self.camera_matrix = None
        self.is_calibrated = False
        self.last_time = 0
        self.current_time = 0
        self.last_time_w = 0
        self.get_logger().info('Node started')

        # self.robot_marker = 1 # синий
        # self.robot_marker = 7 # желтый
        self.robot_marker = 69


    def transform(self, coordinates, z=0):
        # self.get_logger().info(f"Alpha: \n{alpha}")
        center = np.mean(coordinates, axis=0).astype(int)
        center = np.array([center[0], center[1], 1])
        
        alpha = (-z + self.r[-1,:] @ self.t)/(self.r[-1,:] @ center)

        pose = self.r @ (alpha * center - self.t)
        pose[1] = 2000 - pose[1]
        pose /= 1000

        coordinates = np.array(coordinates, dtype=np.float32)
        coordinates = coordinates.reshape(4, 1, 2)
        # print(object_corners.shape, corner_coord.shape)

        retval, rvec, tvec = cv2.solvePnP(self.object_corners, coordinates, self.camera_matrix, self.dist_coeffs, flags=cv2.SOLVEPNP_IPPE_SQUARE)
        # cv2.drawFrameAxes(
        #     self.image, 
        #     self.camera_matrix, 
        #     self.dist_coeffs, 
        #     rvec, 
        #     tvec, 
        #     self.marker_length/2
        # )
        rotation_matrix, _ = cv2.Rodrigues(rvec)
        theta = np.arctan2(rotation_matrix[0,0], rotation_matrix[1,0])

        return pose[:2], theta

    def find_markers(self, calibrate=False):
        marker_corners, marker_IDs, reject = self.detector.detectMarkers(self.image)
        # self.get_logger().info(f"Founded markres with ids: {marker_IDs.ravel()}")
        
        if marker_corners:
            if calibrate:
                return marker_corners, marker_IDs
            else:
                result = {}
                for ids, corners in zip(marker_IDs, marker_corners):
                    corners = corners.reshape(4, 2)
                    corners = corners.astype(int)
                    # top_left = corners[0].ravel()
                    # top_right = corners[1].ravel()
                    bottom_right = corners[2].ravel()
                    # bottom_left = corners[3].ravel()
                    # center_coordinates = (int(top_left[0] + top_right[0]) // 2, int(top_left[1] + top_right[1]) // 2)
                    # self.image = cv2.circle(self.image, center_coordinates, 5, (0, 255, 0), 4)
                    cv2.putText(self.image, str(ids),
                            (bottom_right[0], bottom_right[1] - 15),
                            cv2.FONT_HERSHEY_SIMPLEX,
                            0.5, (0, 255, 0), 2)

                    # center = (int(top_left[0] + bottom_right[0]) // 2, int(top_left[1] + bottom_right[1]) // 2, 1)
                    # center = np.array(center)
                    
                    result[ids[0]] = corners
                    # print(result[ids[0]])
                    # result[ids[0]] = center, np.hstack((top_left, 1))
                return result
        else:
            return None
    
    def calibrate(self):
        marker_corners, marker_IDs = self.find_markers(True)
        objPoints, imgPoints = self.board.matchImagePoints(marker_corners, marker_IDs)
        print(objPoints)
        print(imgPoints)
        retval, rvec, tvec = cv2.solvePnP(objPoints, imgPoints, self.camera_matrix, self.dist_coeffs, flags=cv2.SOLVEPNP_IPPE)
        # self.get_logger().info(f"Tvec: \n{tvec}")
        # self.get_logger().info(f"Rvec: {rvec}")
        # self.get_logger().info(f"Rmat: \n{cv2.Rodrigues(rvec)[0]}")

        # self.image = cv2.drawFrameAxes(self.image, self.camera_matrix, self.dist_coeffs, rvec, tvec, length=1000)

        r_mat = cv2.Rodrigues(rvec)[0]
        Rt = np.hstack((r_mat, tvec))
        T = self.camera_matrix @ Rt
        
        self.r = np.linalg.inv(T[:,:-1])  # final rotate matrix
        self.t = T[:,-1]  # final translate vector

        self.get_logger().info(f"Calibrated succesfuly!")
        self.get_logger().info(f"r: \n{self.r}")
        self.get_logger().info(f"t: {self.t}")

        self.set_default()
        self.is_calibrated = True


    def set_default(self):
        markers = self.find_markers()
        pose, theta = self.transform(markers[self.robot_marker], 435)
        # self.default_theta = 3.874187464676028
        self.default_theta = theta
        self.get_logger().info(f"Default theta: {self.default_theta}")
        self.last_x, self.last_y, self.last_theta = pose[1], -pose[0], 0
        self.get_logger().info(f"Start odometry: {self.last_x, self.last_y, self.last_theta}")
        self.last_time = self.get_clock().now().nanoseconds
        self.last_time_w = self.last_time

    def send_tf(self, x, y, theta):
        t = TransformStamped()

        t.header.stamp = self.get_clock().now().to_msg()
        t.header.frame_id = 'map'
        t.child_frame_id = 'base_link'

        t.transform.translation.x = y
        t.transform.translation.y = -x
        t.transform.translation.z = 0.01

        q = quaternion_from_euler(0, 0, theta)
        # q = quaternion_from_euler(0, 0, -theta)
        # self.get_logger().info(f"Orientation: {q}")
        t.transform.rotation.x = q[0]
        t.transform.rotation.y = q[1]
        t.transform.rotation.z = q[2]
        t.transform.rotation.w = q[3]

        self.tf_broadcaster.sendTransform(t)

    def send_odometry(self, x, y, theta):
        odom = Odometry()

        odom.header.stamp = self.get_clock().now().to_msg()
        odom.header.frame_id = 'map'
        
        # self.get_logger().info(f"Time: {self.get_clock().now().nanoseconds}")

        
        dt = (self.current_time - self.last_time) / 10**9
        dt_w = (self.current_time - self.last_time_w) / 10**9
        # dt = 0.1
        # self.get_logger().info(f"Delta time: {dt}")

        # self.get_logger().info(f"Pose: {y, -x}, theta: {theta}")
        
        vel_x, vel_y, vel_w = (y - self.last_x)/dt, (-x - self.last_y)/dt, (theta - self.last_theta)/dt_w 

        # vel_x, vel_y, vel_w = (y - self.last_x)/dt, (-x - self.last_y)/dt, (-theta - self.last_theta)/dt_w 
        # vel_x, vel_y, vel_w = 0.0, 0.0, 0.0
        # if abs(vel_w) > 1:
        #     self.get_logger().warn(f"W Velosity: {vel_w} {np.degrees(theta):.5f} {np.degrees(-self.last_theta):.5f}")
        #     theta = -self.last_theta
        # else:
        #     self.last_theta = -theta
        #     self.last_time_w = self.current_time

        if abs(vel_w) > 1:
            self.get_logger().warn(f"W Velosity: {vel_w} {np.degrees(theta):.5f} {np.degrees(self.last_theta):.5f}")
            theta = self.last_theta
        else:
            self.last_theta = theta
            self.last_time_w = self.current_time


        # q = quaternion_from_euler(0, 0, -theta)
        q = quaternion_from_euler(0, 0, theta)
        # set the position
        odom.pose.pose.position.x = y
        odom.pose.pose.position.y = -x
        odom.pose.pose.position.z = 0.01;        
        odom.pose.pose.orientation.x = q[0]
        odom.pose.pose.orientation.y = q[1]
        odom.pose.pose.orientation.z = q[2]
        odom.pose.pose.orientation.w = q[3]
        
        self.get_logger().info(f"Odometry: {y:.5f} {-x:.5f} {np.degrees(theta):.5f} {vel_w:.5f}")
        # self.get_logger().info(f"Velosity: {vel_x, vel_y, vel_w}")
        # set the velocity
        odom.child_frame_id = 'base_link'
        odom.twist.twist.linear.x = vel_x
        odom.twist.twist.linear.y = vel_y
        # odom.twist.twist.linear.x = 0.00
        # odom.twist.twist.linear.y = 0.00
        odom.twist.twist.linear.z = 0.00

        odom.twist.twist.angular.x = 0.0
        odom.twist.twist.angular.y = 0.0
        # odom.twist.twist.angular.z = 0.0
        odom.twist.twist.angular.z = vel_w

        self.odom_pub.publish(odom)

        self.end_time = self.get_clock().now().nanoseconds
        
        self.last_x = y
        self.last_y = -x
        self.last_time = self.current_time

    def timer_callback(self):
        self.send_odometry(1.0, 1.0, 0)

    def callback(self, msg):
        try:
            self.image = self.bridge.imgmsg_to_cv2(msg, "bgr8")
        except:
            self.get_logger().warning("No image")

        # FLIP IMAGE FOR REAL 
        # self.image = cv2.flip(self.image, -1)
        self.current_time = self.get_clock().now().nanoseconds

        # np_arr = np.frombuffer(msg.data, np.uint8)
        # self.image = cv2.imdecode(np_arr, cv2.IMREAD_COLOR)
        h,  w = self.image.shape[:2]
        if self.camera_matrix is None:

            # FOR REAL
            # self.camera_matrix = cv2.fisheye.estimateNewCameraMatrixForUndistortRectify(
            #     self.K, self.D, (w, h), np.eye(3), balance=0 # balance=0 (обрезка краёв) ... 1 (сохранение всех пикселей)
            # )
            
            # FOR SIM
            self.camera_matrix = np.array([[761.80910110473633, 0., 960], [0., 761.80913686752319, 540], [0., 0., 1.]], dtype=np.float32)
            
            self.dist_coeffs = np. array([0, 0, 0, 0, 0])

        # FOR REAL
        # undistorted = cv2.fisheye.undistortImage(self.image, self.K, self.D, None, self.camera_matrix)
        # self.image = undistorted
        
        if self.is_calibrated:
            try:
                markers = self.find_markers()
                # self.get_logger().info(f"Founded markres: {markers.keys()}")
                pose, theta = self.transform(markers[self.robot_marker], 435)

                # Публикация трансформа (использовать если камера - единственный источник одометрии)
                # self.send_tf(pose[0], pose[1], theta)

                self.send_odometry(pose[0], pose[1], theta)
            except Exception as e:
                self.get_logger().warning(f"Error finding: {e}")
        else:
            try:
                self.calibrate()

            # # self.get_logger().info(f"Coordinates: {result}")
            except Exception as e:
                self.get_logger().error(f'Error calibrating: {e}')
        # self.get_logger().info(self.get_clock().now().to_msg())
        # self.end_start_timetime = self.get_clock().now().nanoseconds
        # print(self.start_time, self.end_time)
        
        
        # cv2.putText(self.image, "Ping = " + str((self.end_time - self.start_time)//(10**6)) + "ms", (self.image.shape[1] - 400, 30), cv2.FONT_HERSHEY_SIMPLEX, 1, (0,255,0), 2)
        cv2.imshow("Camera Image", self.image)
        key = cv2.waitKey(1)
        if key == ord('c'):
            pass
        if key == ord('q'):
            raise SystemExit


def main(args=None):
    rclpy.init(args=args)
    node = CusCamera()
    try:
        rclpy.spin(node)
    except SystemExit:
        rclpy.logging.get_logger("Quitting").info('Done')
    node.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: src/camera_cus_cpp/src/camera_node.cpp
#include <chrono>
#include <memory>
#include <map>
#include <cmath>
#include <vector>
#include <tuple>
#include <algorithm>

#include "rclcpp/rclcpp.hpp"
#include "sensor_msgs/msg/image.hpp"
#include "nav_msgs/msg/odometry.hpp"
#include "geometry_msgs/msg/transform_stamped.hpp"
#include "tf2_ros/transform_broadcaster.h"
#include "tf2/LinearMath/Quaternion.h"
#include "tf2_geometry_msgs/tf2_geometry_msgs.hpp"
#include "cv_bridge/cv_bridge.hpp"
#include "opencv2/opencv.hpp"
#include "opencv2/aruco.hpp"

using namespace std::chrono_literals;

class CusCamera : public rclcpp::Node
{
public:
    CusCamera() : Node("cus_camera"), is_calibrated_(false), marker_length_(80), robot_marker_id_(69)
    {
        RCLCPP_INFO(this->get_logger(), "OpenCV version: %d", CV_VERSION_MAJOR);

        this->set_parameter(rclcpp::Parameter("use_sim_time", true));

        rclcpp::QoS qos(rclcpp::KeepLast(2));
        qos.reliability(rclcpp::ReliabilityPolicy::BestEffort);
        qos.durability(rclcpp::DurabilityPolicy::Volatile);
        image_sub_ = this->create_subscription<sensor_msgs::msg::Image>(
            "/camera/image", qos, std::bind(&CusCamera::imageCallback, this, std::placeholders::_1));

        odom_pub_ = this->create_publisher<nav_msgs::msg::Odometry>("camera_odom", 1000);
        tf_broadcaster_ = std::make_shared<tf2_ros::TransformBroadcaster>(this);

        // Старый API: словарь и параметры как Ptr
        dictionary_ = cv::aruco::getPredefinedDictionary(cv::aruco::DICT_4X4_250);
        detector_params_ = cv::aruco::DetectorParameters::create();

        int half = marker_length_ / 2;
        object_corners_ = {
            cv::Point3f(-half,  half, 0),
            cv::Point3f( half,  half, 0),
            cv::Point3f( half, -half, 0),
            cv::Point3f(-half, -half, 0)
        };

        // Карта маркеров доски (сохраняем для калибровки)
        board_markers_ = {
            {20, {{550, 550, 0}, {650, 550, 0}, {650, 650, 0}, {550, 650, 0}}},
            {21, {{2350, 550, 0}, {2450, 550, 0}, {2450, 650, 0}, {2350, 650, 0}}},
            {22, {{550, 1350, 0}, {650, 1350, 0}, {650, 1450, 0}, {550, 1450, 0}}},
            {23, {{2350, 1350, 0}, {2450, 1350, 0}, {2450, 1450, 0}, {2350, 1450, 0}}}
        };

        // Создаём объект доски (для совместимости, но не используем matchImagePoints)
        std::vector<std::vector<cv::Point3f>> obj_points;
        std::vector<int> ids;
        for (const auto& m : board_markers_) {
            ids.push_back(m.first);
            obj_points.push_back(m.second);
        }
        cv::aruco::Dictionary board_dict = cv::aruco::getPredefinedDictionary(cv::aruco::DICT_4X4_50);
        auto board_dict_ptr = cv::makePtr<cv::aruco::Dictionary>(board_dict);
        board_ = cv::aruco::Board::create(obj_points, board_dict_ptr, ids);

        // Матрица камеры и коэффициенты искажений (для симуляции)
        camera_matrix_ = (cv::Mat_<double>(3, 3) << 761.80910110473633, 0., 960.,
                                                      0., 761.80913686752319, 540.,
                                                      0., 0., 1.);
        dist_coeffs_ = cv::Mat::zeros(5, 1, CV_64F);

        RCLCPP_INFO(this->get_logger(), "Node started");
    }

private:
    static geometry_msgs::msg::Quaternion quaternionFromEuler(double roll, double pitch, double yaw)
    {
        tf2::Quaternion q;
        q.setRPY(roll, pitch, yaw);
        return tf2::toMsg(q);
    }

    std::map<int, std::vector<cv::Point2f>> findMarkers(bool calibrate = false)
    {
        std::map<int, std::vector<cv::Point2f>> result;
        std::vector<std::vector<cv::Point2f>> corners;
        std::vector<int> ids;
        std::vector<std::vector<cv::Point2f>> rejected;

        // Используем старый функциональный API
        cv::aruco::detectMarkers(image_, dictionary_, corners, ids, detector_params_, rejected);

        if (!ids.empty()) {
            for (size_t i = 0; i < ids.size(); ++i) {
                if (calibrate) {
                    result[ids[i]] = corners[i];
                } else {
                    cv::putText(image_, std::to_string(ids[i]),
                                corners[i][2],
                                cv::FONT_HERSHEY_SIMPLEX, 0.5, cv::Scalar(0, 255, 0), 2);
                    result[ids[i]] = corners[i];
                }
            }
        }
        return result;
    }

    void calibrate()
    {
        auto markers = findMarkers(true);
        if (markers.empty()) {
            RCLCPP_WARN(this->get_logger(), "No markers found for calibration");
            return;
        }

        // Собираем все 3D и 2D точки всех маркеров доски
        std::vector<cv::Point3f> all_obj_points;
        std::vector<cv::Point2f> all_img_points;

        for (const auto& m : markers) {
            int id = m.first;
            const auto& img_corners = m.second;
            auto it = board_markers_.find(id);
            if (it != board_markers_.end()) {
                const auto& obj_corners = it->second;
                for (size_t i = 0; i < img_corners.size(); ++i) {
                    all_obj_points.push_back(obj_corners[i]);
                    all_img_points.push_back(img_corners[i]);
                }
            }
        }

        if (all_obj_points.empty()) {
            RCLCPP_ERROR(this->get_logger(), "Failed to match board points");
            return;
        }

        cv::Mat rvec, tvec;
        bool success = cv::solvePnP(all_obj_points, all_img_points, camera_matrix_, dist_coeffs_,
                                    rvec, tvec, false, cv::SOLVEPNP_IPPE);
        if (!success) {
            RCLCPP_ERROR(this->get_logger(), "solvePnP failed during calibration");
            return;
        }

        cv::Mat R_board;
        cv::Rodrigues(rvec, R_board);
        cv::Mat Rt(3, 4, CV_64F);
        R_board.copyTo(Rt(cv::Rect(0, 0, 3, 3)));
        tvec.copyTo(Rt(cv::Rect(3, 0, 1, 3)));

        cv::Mat T = camera_matrix_ * Rt;
        cv::Mat T_33 = T(cv::Rect(0, 0, 3, 3));
        cv::invert(T_33, r_);
        t_ = T.col(3);

        RCLCPP_INFO(this->get_logger(), "Calibration successful");
        RCLCPP_INFO_STREAM(this->get_logger(), "r:\n" << r_);
        RCLCPP_INFO_STREAM(this->get_logger(), "t: " << t_.t());

        setDefault();
        is_calibrated_ = true;
    }

    void setDefault()
    {
        auto markers = findMarkers();
        if (markers.find(robot_marker_id_) == markers.end()) {
            RCLCPP_ERROR(this->get_logger(), "Robot marker not found during setDefault");
            return;
        }
        double x, y, theta;
        std::tie(x, y, theta) = transform(markers[robot_marker_id_], 435.0);
        default_theta_ = theta;
        last_x_ = y;
        last_y_ = -x;
        last_theta_ = 0.0;
        last_time_ = this->now();
        last_time_w_ = last_time_;
        RCLCPP_INFO(this->get_logger(), "Default theta: %.5f", default_theta_);
        RCLCPP_INFO(this->get_logger(), "Start odometry: (%.5f, %.5f, %.5f)",
                    last_x_, last_y_, last_theta_);
    }

    std::tuple<double, double, double> transform(const std::vector<cv::Point2f>& corners, double z)
    {
        // Центр маркера в пикселях
        cv::Point2f center_px(0, 0);
        for (const auto& p : corners) {
            center_px.x += p.x;
            center_px.y += p.y;
        }
        center_px.x /= 4.0;
        center_px.y /= 4.0;
        cv::Mat center = (cv::Mat_<double>(3, 1) << center_px.x, center_px.y, 1.0);

        cv::Mat r_last_row = r_.row(2);
        double alpha_num = -z + r_last_row.dot(t_);
        double alpha_den = r_last_row.dot(center);
        double alpha = alpha_num / alpha_den;

        cv::Mat pose = r_ * (alpha * center - t_);
        double x = pose.at<double>(0);
        double y = pose.at<double>(1);

        y = 2000.0 - y;
        x /= 1000.0;
        y /= 1000.0;

        // Оценка поворота маркера
        std::vector<cv::Point3f> obj_pts = object_corners_;
        std::vector<cv::Point2f> img_pts = corners;
        cv::Mat rvec_marker, tvec_marker;
        cv::solvePnP(obj_pts, img_pts, camera_matrix_, dist_coeffs_,
                     rvec_marker, tvec_marker, false, cv::SOLVEPNP_IPPE_SQUARE);
        cv::Mat R_marker;
        cv::Rodrigues(rvec_marker, R_marker);
        double theta = std::atan2(R_marker.at<double>(0, 0), R_marker.at<double>(1, 0));

        return {x, y, theta};
    }

    void sendOdometry(double x, double y, double theta)
    {
        nav_msgs::msg::Odometry odom;
        odom.header.stamp = this->now();
        odom.header.frame_id = "map";
        odom.child_frame_id = "base_link";

        double dt = (this->now() - last_time_).seconds();
        double dt_w = (this->now() - last_time_w_).seconds();

        double vel_x = (y - last_x_) / dt;
        double vel_y = (-x - last_y_) / dt;
        double vel_w = (theta - last_theta_) / dt_w;

        if (std::abs(vel_w) > 1.0) {
            RCLCPP_WARN(this->get_logger(), "High angular velocity: %.5f rad/s", vel_w);
            theta = last_theta_;
        } else {
            last_theta_ = theta;
            last_time_w_ = this->now();
        }

        odom.pose.pose.position.x = y;
        odom.pose.pose.position.y = -x;
        odom.pose.pose.position.z = 0.01;
        odom.pose.pose.orientation = quaternionFromEuler(0, 0, theta);

        odom.twist.twist.linear.x = vel_x;
        odom.twist.twist.linear.y = vel_y;
        odom.twist.twist.angular.z = vel_w;

        odom_pub_->publish(odom);

        RCLCPP_INFO(this->get_logger(), "Odometry: x=%.5f y=%.5f theta=%.5f deg, vel_w=%.5f",
                    y, -x, theta * 180.0 / M_PI, vel_w);

        last_x_ = y;
        last_y_ = -x;
        last_time_ = this->now();
    }

    void imageCallback(const sensor_msgs::msg::Image::SharedPtr msg)
    {
        try {
            auto cv_ptr = cv_bridge::toCvCopy(msg, "bgr8");
            image_ = cv_ptr->image;
        } catch (cv_bridge::Exception& e) {
            RCLCPP_WARN(this->get_logger(), "cv_bridge exception: %s", e.what());
            return;
        }

        if (!is_calibrated_) {
            calibrate();
        } else {
            try {
                auto markers = findMarkers();
                if (markers.find(robot_marker_id_) != markers.end()) {
                    auto [x, y, theta] = transform(markers[robot_marker_id_], 435.0);
                    sendOdometry(x, y, theta);
                } else {
                    RCLCPP_WARN_THROTTLE(this->get_logger(), *this->get_clock(), 3000,
                                         "Robot marker %d not found", robot_marker_id_);
                }
            } catch (const std::exception& e) {
                RCLCPP_WARN(this->get_logger(), "Error during processing: %s", e.what());
            }
        }

        cv::imshow("Camera Image", image_);
        int key = cv::waitKey(1);
        if (key == 'q') {
            rclcpp::shutdown();
        }
    }

    // Члены класса
    rclcpp::Subscription<sensor_msgs::msg::Image>::SharedPtr image_sub_;
    rclcpp::Publisher<nav_msgs::msg::Odometry>::SharedPtr odom_pub_;
    std::shared_ptr<tf2_ros::TransformBroadcaster> tf_broadcaster_;

    cv::Mat image_;
    cv::Ptr<cv::aruco::Dictionary> dictionary_;
    cv::Ptr<cv::aruco::DetectorParameters> detector_params_;
    cv::Ptr<cv::aruco::Board> board_;
    std::vector<cv::Point3f> object_corners_;

    cv::Mat camera_matrix_, dist_coeffs_;
    bool is_calibrated_;
    cv::Mat r_, t_;
    double default_theta_;

    double last_x_, last_y_, last_theta_;
    rclcpp::Time last_time_, last_time_w_;

    const int marker_length_;
    const int robot_marker_id_;

    std::map<int, std::vector<cv::Point3f>> board_markers_;
};

int main(int argc, char** argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<CusCamera>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}